### DISCUSSION TITLE
Two Optimization opinion

### DISCUSSION THREAD

**Discussion by SnowWindSaveYou** [happy-turtle/oit-unity#8] [Ideas]
just fixed two main performance issues, one is the "SetData" in the PreRender, another is the "Blit" in the Render.

for the first one, i use a compute shader replace the method of set value from cpu to gpu. 

`OrderIndependentTransparency/Shaders/Resources/OitComputeUtils.compute`
```
#pragma kernel ClearStartOffsetBuffer

RWStructuredBuffer<uint> StartOffsetBuffer;

[numthreads(64,1,1)]
void ClearStartOffsetBuffer (uint3 id : SV_DispatchThreadID)
{
    StartOffsetBuffer[id.x] = 0;
}

```

```c#

public class OitLinkedList : IOrderIndependentTransparency
{
   ...
    private ComputeShader OitComputeUtils;
    private int clearStartOffsetBufferKernel;
    private int bufferSizeHead = 0;

    public OitLinkedList(bool postProcess = false)
    {
       ...
        OitComputeUtils = Resources.Load<ComputeShader>("OitComputeUtils");
        clearStartOffsetBufferKernel = OitComputeUtils.FindKernel("ClearStartOffsetBuffer");
        OitComputeUtils.SetBuffer(clearStartOffsetBufferKernel,startOffsetBufferId, startOffsetBuffer);
    }

    public void PreRender()
    {
        if (fragmentLinkBuffer == null || startOffsetBuffer == null)
            return;

        //reset StartOffsetBuffer to zeros

        // set buffers for rendering
        Graphics.SetRandomWriteTarget(1, fragmentLinkBuffer);
        Graphics.SetRandomWriteTarget(2, startOffsetBuffer);
        OitComputeUtils.Dispatch(clearStartOffsetBufferKernel, bufferSizeHead / 64, 1, 1);
        //startOffsetBuffer.SetData(resetTable);
    }
}
```

For the second one, i'm render the result on a quad mesh to avoid the Blit ,but my implementation is inside my original volume rendering shader.
it can be done by DrawProcedual ,but i have no idea about how to remove the Blit from post process renderer.

PS:
i tried real insert sort for reduce the number of swap times, but no mush performance difference 

```
nNumPixels = min(MAX_SORTED_PIXELS,nNumPixels);

for (int ii = 0; ii < nNumPixels-1; ii++)
{
      float minDepth = UnpackDepth(SortedPixels[ii].uDepthCoverage);
      int minIdx = ii;
  
      for (int j = ii+1;j<nNumPixels;j++)
      {
      float depth = UnpackDepth(SortedPixels[j].uDepthCoverage);//linear01Depth
              minDepth = min(depth,minDepth);
              minIdx = lerp( minIdx,j,minDepth==depth);
      }
      FragmentAndLinkBuffer_STRUCT temp = SortedPixels[ii];
      SortedPixels[ii] = SortedPixels[minIdx];
      SortedPixels[minIdx] = temp;
}
```

**Comment by happy-turtle** [comment]
Very cool ideas! Did you measure performance with the Unity profiler? I will try to test your suggestions. Maybe I have an idea for the post process, it is much more ergonomic as a post process I think.
I have no objections against a compute shader replacing the SetData method. We require a shader level where compute shaders should be available anyways. If you like you could open a pull request for it.

I tried insert sort at some point as well, but disregarded it because at that time I didn't want to add extra complexity. And I messed it up a bit I think. Cool thing you were able to test it! If it proves useful at some point we could add it as well.

**Reply by SnowWindSaveYou** [reply]
ok, i opened a pull req for it.
Actually for the virtual reality projects I'm working on, the difference of performance between these methods is very clear, the performance of setData is related to screen size, and the VR got relative larger screen and more render count, that If I don't fix these issues, I would only get about 60FPS. 

but change the SetData to the compute shader also leads other problem, it can't work properly on a very large screen (e.g. 4K HUD 3840*2160) due to the buffer counts over the max thread group. while it can be fixed by split thread size into many dispatches with an index offset.

**Comment by happy-turtle** [comment]
> For the second one, i'm render the result on a quad mesh to avoid the Blit ,but my implementation is inside my original volume rendering shader. it can be done by DrawProcedual ,but i have no idea about how to remove the Blit from post process renderer.
> 

You can do exactly the same in the post process renderer. You can call `context.command.DrawProcedural(...)` instead of `context.command.Blit(...)`. What parameters do you use to draw the fullscreen triangle? Sounds like a good improvement as well!

**Reply by happy-turtle** [reply]
The [docs](https://docs.unity3d.com/ScriptReference/Graphics.Blit.html) say Blit actually also remders a fullscreen quad. Maybe need some testing if it actually makes a difference compared to DrawProcedural?
